### PR TITLE
Implement SVM J2I thunk validation and guarantee resolved virtual dispatch under SVM on x86

### PR DESCRIPTION
**Pull request by jdmpapin**
The J2I thunk from method validation ensures that the J2I thunk appropriate to the method specified by the given method ID exists and is registered with the VM, and uses the address of the J2I thunk to define the given thunk ID.

Then with this validation, it's very straightforward to guarantee resolved virtual dispatch under SVM on x86. This will allow the IL generator to generate a resolved virtual call node for `invokeinterface` when calling a non-final method of `Object`, e.g. `toString()`.

This PR requires eclipse/omr#6358 to be merged first

### CODE REVIEW

**Comment by jdmpapin** [comment]
It's likely that a corresponding change should be just as straightforward in other code generators, but those changes are not included in this PR

**Review by dsouzai — CHANGES_REQUESTED** [review]
Overall LGTM. However, there is one change requested regarding the `pthread_jit_write_protect_np` calls that are very likely unnecessary or even potentially problematic.

File: runtime/compiler/runtime/RelocationRecord.cpp [hunk 1]
```diff
@@ -2402,11 +2429,19 @@ TR_RelocationRecordJ2IVirtualThunkPointer::relocateJ2IVirtualThunkPointer(
    {
    TR_ASSERT_FATAL(thunk != NULL, "expected a j2i virtual thunk for relocation\n");
 
+#if defined(OSX) && defined(AARCH64)
+   pthread_jit_write_protect_np(0);
+#endif
+
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/runtime/RelocationRecord.cpp:2435]: @0xdaryl could you provide feedback regarding this as it is my understanding you're involved in the M1 OSX related changes.

**Comment by dsouzai** [comment]
Oh also, you need to update the minor version for JITServer, as a new relocation record has been added (see https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/BestPractices.md#jitserver)

File: runtime/compiler/runtime/RelocationRecord.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xdaryl) [reply runtime/compiler/runtime/RelocationRecord.cpp:2435]: This pair of functions is needed to guard writes to executable memory.  Hence, you only need to remove execute permission and then restore it for writes to code that will actually be executed (i.e., in the code cache).  It isn't clear to me whether this relocation is patching the code cache or not.  But assuming it is, you're right that these calls don't need to be here if it is guaranteed that there will always be an outer enclosing pair.

One thing to consider is if this function would ever be called outside of that protective wrapper.  If that is the case, then you should probably keep the API calls here.

[the remark this answers is not in the document]
Comment (jdmpapin) [reply runtime/compiler/runtime/RelocationRecord.cpp:2435]: > Hence, you only need to remove execute permission and then restore it for writes to code that will actually be executed (i.e., in the code cache)

This method isn't patching code that will be executed per se, but the data that it is patching is within the code cache. All current uses of this relocation are for the data portion of VPIC and IPIC snippets

The other occurrence in the static `relocateAndRegisterThunk()` is just something I preserved from the body of `TR_RelocationRecordThunks::relocateAndRegisterThunk()`. In that case, we're copying the body of a J2I thunk into the code cache, and that is code that may be executed later

So the code cache must be writable in both of these paths. However, I don't anticipate ever needing to call either of these outside of `applyRelocationAtAllOffsets()`, so I'll remove these calls and rely on write protection having been set appropriately further up the call stack

**Comment by jdmpapin** [comment]
Removed the `pthread_jit_write_protect_np()` calls and bumped JITServer's `MINOR_NUMBER`

**Comment by jdmpapin** [comment]
Rebased to fix the `MINOR_NUMBER` conflict

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
jenkins test sanity.functional+aot xlinux jdk17

**Comment by dsouzai** [comment]
jenkins test sanity all jdk17

**Comment by dsouzai** [comment]
jenkins test sanity all jdk17

**Comment by dsouzai** [comment]
s390x failure because of https://github.com/eclipse-openj9/openj9/issues/13128

win64 failure seems to be infra related?

**Comment by dsouzai** [comment]
jenkins test sanity.functional+aot xlinux jdk17

**Comment by dsouzai** [comment]
The AOT build was aborted because I guess it took too long to get scheduled/run -_-

**Comment by dsouzai** [comment]
jenkins test sanity.functional+aot xlinux jdk17

**Comment by jdmpapin** [comment]
@IBMJimmyk, since we discussed doing the same change on Power, see the final (x86) commit in this PR. If Power code gen already only generates a plain call through the vtable for resolved virtual calls, it should be similar

@joransiu, I haven't discussed this with you, but a similar statement applies to z in case you want to improve code generation for this case in SVM-based AOT

**Comment by knn-k** [comment]
I opened #14638 to fix build failures on AArch64 macOS.
You need to call `pthread_jit_write_protect_np()` before and after writing to the JIT code cache.

**Pull request merged**